README.md
[![npm version](https://badge.fury.io/js/%40aws-lambda%2Fhttp.svg)](https://badge.fury.io/js/%40aws-lambda%2Fhttp)
[![build status](https://github.com/bravecloud/aws-http-lambda/actions/workflows/build.yml/badge.svg?branch=master)](https://github.com/bravecloud/aws-http-lambda/actions/workflows/build.yml)
[![Coverage Status](https://coveralls.io/repos/github/bravecloud/aws-http-lambda/badge.svg?branch=master)](https://coveralls.io/github/bravecloud/aws-http-lambda?branch=master)

# @aws-lambda/http

http is a utility for simplifying response handling when using aws lambda functions with the http event proxy.

## Table of Contents

1. [Installation and Usage](#installation-and-usage)
2. [Examples](#examples)


## <a name="installation-and-usage"></a>Installation and Usage

Prerequisites: [Node.js](https://nodejs.org/) `>=8`, npm version 6+.

You can install the http module using npm:

```
$ npm install @aws-lambda/http --save
```

## <a name="examples"></a>Response Examples

### AWS Lambda handler response with JSON body and header

```javascript
const Response = require("@aws-lambda/http").Response;

module.exports.hello = async event => {
  return new Response()
                .status(200)
                .body({"message" : "hello world!"})
                .header("x-api-version", "1.0")
                .json();
};
```

### Adding HTTP security headers to response (alternative to [Lambda@Edge + CloudFront](https://aws.amazon.com/blogs/networking-and-content-delivery/adding-http-security-headers-using-lambdaedge-and-amazon-cloudfront/))

```javascript
const HTTP = require("@aws-lambda/http");
const Response = HTTP.Response;

module.exports.hello = async event => {
  return new Response()
                .status(200)
                .body({"message" : "hello world!"})
                .header("x-api-version", "1.0")
                .headers(HTTP.SecurityHeaders)
                .json();
};
```

### Adding cookie to response

```javascript
const HTTP = require("@aws-lambda/http");
const Response = HTTP.Response;
const Cookie = HTTP.Cookie;

module.exports.hello = async event => {
  return new Response()
                .status(200)
                .body({"message" : "hello world!"})
                .header("x-api-version", "1.0")
                .headers(HTTP.SecurityHeaders)
                .cookie(new Cookie("SSID", "ubmHxGlgi71l8ayOMw9f6wknjGv2FwDKLt")))
                .json();
};
```

### Adding secure cookie to response

```javascript
const HTTP = require("@aws-lambda/http");
const Response = HTTP.Response;
const Cookie = HTTP.Cookie;

module.exports.hello = async event => {
  return new Response()
                .status(200)
                .body({"message" : "hello world!"})
                .header("x-api-version", "1.0")
                .headers(HTTP.SecurityHeaders)
                .cookie(Cookie.Secure("SSID", "ubmHxGlgi71l8ayOMw9f6wknjGv2FwDKLt"))
                .json();
};
```

### Adding multiple cookies to response

```javascript
const HTTP = require("@aws-lambda/http");
const Response = HTTP.Response;
const Cookie = HTTP.Cookie;

module.exports.hello = async event => {
  return new Response()
                .status(200)
                .body({"message" : "hello world!"})
                .header("x-api-version", "1.0")
                .headers(HTTP.SecurityHeaders)
                .cookies([
                  Cookie.Secure("SSID", "ubmHxGlgi71l8ayOMw9f6wknjGv2FwDKLt"),
                  Cookie.Secure("secret", "shhhhh").expireInMinutes(15),
                  new Cookie("lang", "en").path("/docs")
                ])
                .json();
};
```


CHANGELOG.md
# Changelog

## [1.0.1 - 2022-04-17]

- fixing github actions badge link

## [1.0.0 - 2022-04-10]

- migrating from travis ci to github actions
- updating dependencies


test/testCookie.js
const Cookie = require("../lib/Cookie");

const assert = require('assert');

const UTC_STRING_RE = /Expires=([a-zA-Z]{3},\s\d{1,2}\s[a-zA-z]{3}\s\d{4}\s\d{1,2}:\d{1,2}:\d{1,2}\s[a-zA-Z]{3});/gi

describe('Cookie', function () {
  it('Instantiation', function () {
    let cookie = new Cookie();
  });
  it('SameSite Static Constants', function () {
    assert(Cookie.SS_LAX !== undefined);
    assert(Cookie.SS_STRICT !== undefined);
    assert(Cookie.SS_NONE !== undefined);
  });
  it('Expires', function () {
    let now = new Date();
    let cookie = new Cookie("name", "value");
    let dateStr = "Thu, 18 Jul 2019 01:54:23 GMT";
    assert(cookie.expires(dateStr).toString().indexOf(`Expires=${dateStr}`) !== -1);
  });
  it('Expire in Days', function () {
    let now = new Date();
    let cookie = new Cookie("name", "value");

    let cookieStr = cookie.expireInDays(1).toString();
    let expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60/60/24) === 1);

    cookieStr = cookie.expireInDays(365).toString();
    expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60/60/24) === 365);
  });
  it('Expire in Hours', function () {
    let now = new Date();
    let cookie = new Cookie("name", "value");

    let cookieStr = cookie.expireInHours(1).toString();
    let expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60/60) === 1);

    cookieStr = cookie.expireInHours(12).toString();
    expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60/60) === 12);
  });
  it('Expire in Minutes', function () {
    let now = new Date();
    let cookie = new Cookie("name", "value");

    let cookieStr = cookie.expireInMinutes(1).toString();
    let expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60) === 1);

    cookieStr = cookie.expireInMinutes(30).toString();
    expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000/60) === 30);
  });
  it('Expires in Seconds', function () {
    let now = new Date();
    let cookie = new Cookie("name", "value");

    let cookieStr = cookie.expireInSeconds(1).toString();
    let expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000) === 1);

    cookieStr = cookie.expireInSeconds(30).toString();
    expiresDate = new Date(new RegExp(UTC_STRING_RE).exec(cookieStr)[1]);
    assert(Math.ceil((expiresDate - now.getTime())/1000) === 30);
  });
  it('Basic name + value', function () {
    let cookie = new Cookie("some-name", "some-value");

    let cookieStr = cookie.toString();
    assert(cookieStr.indexOf("some-name") !== -1);
    assert(cookieStr.indexOf("some-value") !== -1);
  });
  it('Max-Age', function () {
    let cookie = new Cookie("some-name", "some-value");
    let maxAge = new Date().getTime() + (1000 * 60 * 5);
    let cookieStr = cookie.maxAge(maxAge).toString();

    assert(cookieStr.toLowerCase().indexOf(`max-age=${maxAge}`) !== -1);
  });
  it('Domain', function () {
    let cookie = new Cookie("some-name", "some-value");
    let domain = "mysite.com";
    let cookieStr = cookie.domain(domain).toString();

    assert(cookieStr.toLowerCase().indexOf(`domain=${domain}`) !== -1);
  });
  it('Path', function () {
    let cookie = new Cookie("some-name", "some-value");
    let path = "/private/images";
    let cookieStr = cookie.path(path).toString();

    assert(cookieStr.toLowerCase().indexOf(`path=${path}`) !== -1);
  });
  it('Secure', function () {
    let cookie = new Cookie("some-name", "some-value");

    assert(cookie.secure().toString().toLowerCase().indexOf("secure") !== -1);
  });
  it('HTTP Only', function () {
    let cookie = new Cookie("some-name", "some-value");

    assert(cookie.httpOnly().toString().toLowerCase().indexOf("httponly") !== -1);
  });
  it('SameSite', function () {
    let cookie = new Cookie("some-name", "some-value");

    assert(cookie.sameSite(Cookie.SS_STRICT).toString().toLowerCase().indexOf("samesite=strict") !== -1);
  });
  it('Secure Cookie Factory Function', function () {
    let cookieStr = Cookie.Secure("some-name", "some-value").toString().toLowerCase();

    assert(cookieStr.indexOf("path=/") !== -1);
    assert(cookieStr.indexOf("secure") !== -1);
    assert(cookieStr.indexOf("httponly") !== -1);
    assert(cookieStr.indexOf("samesite=strict") !== -1);
  });
});

test/testResponse.js
const Response = require("../lib/Response");
const Cookie = require("../lib/Cookie");

const assert = require('assert');
describe('Response', function () {
  it('Instantiation', function () {
    let response = new Response();
  }); 
  it('Status Code', function () {
    let response = new Response();
    //default code
    assert(response.json().statusCode === 200);
    assert(response.status(302).json().statusCode === 302);
  }); 
  it('Headers', function () {
    let response = new Response();
    response
      .header("h1", "v1")
      .header("h2", "v2")
      .header("h3", "v3");
    let built = response.json();
    assert(built.headers["h1"] === "v1");
    assert(built.headers["h2"] === "v2");
    assert(built.headers["h3"] === "v3");
  });
  it('Headers - adding multiple headers', function () {
    let response = new Response();
    response
      .header("h1", "v1")
      .header("h2", "v2")
      .header("h3", "v3");
    response.headers({
      "h4": "v4",
      "h5": "v5"
    });
    let built = response.json();
    assert(built.headers["h1"] === "v1");
    assert(built.headers["h2"] === "v2");
    assert(built.headers["h3"] === "v3");
    assert(built.headers["h4"] === "v4");
    assert(built.headers["h5"] === "v5");
  });
  it('Multi Value Headers', function () {
    let response = new Response();
    response
      .multiValueHeader("h1", ["v1", "v2", "v3"])
      .multiValueHeader("h2", ["v4", "v5", "v6"])
      .multiValueHeader("h3", ["v7", "v8", "v9"]);
    let built = response.json();
    assert(built.multiValueHeaders["h1"].toString() === ["v1", "v2", "v3"].toString());
    assert(built.multiValueHeaders["h2"].toString() === ["v4", "v5", "v6"].toString());
    assert(built.multiValueHeaders["h3"].toString() === ["v7", "v8", "v9"].toString());
  });
  it('Body', function () {
    let response = new Response();
    let jsonBody = {
      "users": [
        {
          "id": 1,   
          "username": "Alice"
        },
        {
          "id": 2,   
          "username": "Body"
        }
      ]
    };
    response.body(jsonBody);
    let built = response.json();
    assert(built.body === JSON.stringify(jsonBody));
    
    let strBody = "Some string message for the body";
    response.body(strBody);
    built = response.json();
    assert(built.body === strBody);

    let numBody= 12345;
    response.body(numBody);
    built = response.json();
    assert(built.body === numBody);
  });
  it('Cookie(s)', function () {
    let response = new Response();
    response.cookie(new Cookie("c1", "v1"));
    let setCookieKey = Object.keys(response.json().multiValueHeaders).filter(key => key.toLowerCase() === "set-cookie")[0];
    assert(setCookieKey);
    assert(response.json().multiValueHeaders[setCookieKey][0] === 'c1=v1;')

    response.cookie(new Cookie("c2", "v2"));
    response.cookie(new Cookie("c3", "v3"));
    response.cookies([
      new Cookie("c4", "v4"),
      new Cookie("c5", "v5"),
      new Cookie("c6", "v6")
    ]);

    let allCookies = response.json().multiValueHeaders[setCookieKey].reduce((acc, curr) => {
      return acc + curr;
    });
    for (let i = 1; i < response.json().multiValueHeaders[setCookieKey].length; i++) {
      let expectedCookie = `c${i}=v${i}`;
      assert(allCookies.indexOf(expectedCookie) !== -1, `missing cookie ${expectedCookie}`);
    }
  });
  it('Cookies added after existing multiheader', function () {
    let response = new Response();
    response.multiValueHeader("Set-Cookie", ["c1=v1;"]);
    response.cookie(new Cookie("c2", "v2"));
    response.cookies([
      new Cookie("c3", "v3"),
      new Cookie("c4", "v4"),
    ]);

    let setCookieKey = Object.keys(response.json().multiValueHeaders).filter(key => key.toLowerCase() === "set-cookie")[0];
    let allCookies = response.json().multiValueHeaders[setCookieKey].reduce((acc, curr) => {
      return acc + curr;
    });

    for (let i = 1; i < response.json().multiValueHeaders[setCookieKey].length; i++) {
      let expectedCookie = `c${i}=v${i}`;
      assert(allCookies.indexOf(expectedCookie) !== -1, `missing cookie ${expectedCookie}`);
    }
  });
  it('isBase64Encoded', function () {
    let response = new Response();
    let json = response.base64().json();
    assert(json.isBase64Encoded);
  });
});

lib/Response.js
const SET_COOKIE = "Set-Cookie";

module.exports = class {
  constructor() {
    this._status = 200;
    this._base64 = undefined;
    this._headers = {};
    this._multiValueHeaders = {};
    this._cookies = [];
    this._body = undefined;
  }

  status(code) {
    this._status = code;
    return this;
  }

  header(name, value) {
    this._headers[name] = value;
    return this;
  }

  headers(headers) {
    this._headers = Object.assign(headers, this._headers);
    return this;
  }

  multiValueHeader(name, values) {
    this._multiValueHeaders[name] = values;
    return this;
  }

  body(value) {
    this._body = value;
    return this;
  }

  cookie(cookie) {
    this._cookies.push(cookie);
    return this;
  }

  cookies(cookies) {
    this._cookies.push(...cookies);
    return this;
  }

  base64() {
    this._base64 = true;
    return this;
  }

  json() {
    //initialize with mandatory values
    let response = {
      statusCode: this._status
    };

    if (this._base64) {
      response.isBase64Encoded = true;
    }

    if (Object.keys(this._headers).length) {
      response.headers = Object.assign({}, this._headers);
    }

    if (Object.keys(this._multiValueHeaders).length) {
      response.multiValueHeaders = {};
      //Can't use Object.assign due to shallow copy limitation, so manually copy key and arrays
      Object.keys(this._multiValueHeaders).forEach(key => {
        response.multiValueHeaders[key] = this._multiValueHeaders[key].slice(0);
      });
    }

    if (this._cookies.length) {
      //Since cookies can manually be set via set-cookie, it's important to check if the header key
      //already exists.
      let setCookieKey = Object.keys(this._multiValueHeaders).filter(key => key.toLowerCase() === SET_COOKIE.toLowerCase());
      //If it already exists, append the cookie values to the multi header value array
      if (setCookieKey.length) {
        response.multiValueHeaders[setCookieKey[0]].push(...this._cookies.map(cookie => cookie.toString()));
      }
      //Otherwise 
      else {
        response.multiValueHeaders = {
          [SET_COOKIE]: this._cookies.map(cookie => cookie.toString())
        };
      }
    }

    if (this._body) {
      response.body = (typeof(this._body) === 'object') ? JSON.stringify(this._body) : this._body;
    }

    return response;
  }
};

lib/Cookie.js
const SECURE = 'Secure';
const HTTP_ONLY = 'HttpOnly';

const MS_IN_SEC = 1000;
const SEC_IN_MIN = 60;
const MIN_IN_HR = 60;
const HR_IN_DAY = 24;

const now = () => {
  return new Date().getTime();
};

module.exports = class {
  static get SS_STRICT() { return "Strict"; }
  static get SS_LAX() { return "Lax"; }
  static get SS_NONE() { return "None"; }

  static Secure(name, value) {
    const Cookie = module.exports;
    return new Cookie(name, value)
                  .secure()
                  .httpOnly()
                  .path("/")
                  .sameSite(Cookie.SS_STRICT);
  }

  constructor(name, value) {
    this._name = name;
    this._value = value;
    this._expires = undefined;
    this._maxAge = undefined;
    this._domain = undefined;
    this._path = undefined;
    this._secure = undefined;
    this._httpOnly = undefined;
    this._sameSite = undefined;
  }

  expires(dateStr) {
    this._expires = dateStr;
    return this;
  }

  expireInDays(numDays) {
    this._expires = new Date(now() + (MS_IN_SEC * SEC_IN_MIN * MIN_IN_HR * HR_IN_DAY * numDays)).toUTCString();
    return this;
  }

  expireInHours(numHours) {
    this._expires = new Date(now() + (MS_IN_SEC * SEC_IN_MIN * MIN_IN_HR * numHours)).toUTCString();
    return this;
  }

  expireInMinutes(numMins) {
    this._expires = new Date(now() + (MS_IN_SEC * SEC_IN_MIN * numMins)).toUTCString();
    return this;
  }

  expireInSeconds(numSecs) {
    this._expires = new Date(now() + (MS_IN_SEC * numSecs)).toUTCString();
    return this;
  }

  maxAge(age) {
    this._maxAge = age;
    return this;
  }

  domain(domain) {
    this._domain = domain;
    return this;
  }

  path(path) {
    this._path = path;
    return this;
  }

  secure() {
    this._secure = SECURE;
    return this;
  }

  httpOnly() {
    this._httpOnly = HTTP_ONLY;
    return this;
  }

  sameSite(policy) {
    this._sameSite = policy;
    return this;
  }

  toString() {
    let cookieStr = `${this._name}=${this._value};`;

    if (this._expires) {
      cookieStr += ` Expires=${this._expires};`;
    }

    if (this._maxAge) {
      cookieStr += ` Max-Age=${this._maxAge};`;
    }

    if (this._domain) {
      cookieStr += ` Domain=${this._domain};`;
    }

    if (this._path) {
      cookieStr += ` Path=${this._path};`;
    }

    if (this._secure) {
      cookieStr += ` ${this._secure};`;
    }

    if (this._httpOnly) {
      cookieStr += ` ${this._httpOnly};`;
    }

    if (this._sameSite) {
      cookieStr += ` SameSite=${this._sameSite};`;
    }
    
    return cookieStr;
  }
};